File: 0079. Word Search.java
// Medium
// DFS
// Time:O(m*n* 4^L), Space:O(L)
// https://leetcode.cn/problems/word-search/

class Solution {
    // DFS
    public boolean exist(char[][] board, String word) {
        int r = board.length, c = board[0].length;
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if (dfs(board, word, i, j, 0)) return true;
            }
        }
        return false;
    }

    private boolean dfs(char[][] board, String word, int i, int j, int index) {
        if (index == word.length()) return true;
        // check boundary
        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length || board[i][j] != word.charAt(index)) return false;

        char tmp = board[i][j];
        board[i][j] = '#';

        boolean found = dfs(board, word, i - 1, j, index + 1) ||
                        dfs(board, word, i + 1, j, index + 1) ||
                        dfs(board, word, i, j - 1, index + 1) ||
                        dfs(board, word, i, j + 1, index + 1);
        
        board[i][j] = tmp;
        return found;
    }
}

File: 0001-0500/0079. Word Search.py
# Time:O(m*n* 4^L), Space:O(L)

from typing import List

class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m, n = len(board), len(board[0])

        # from (i, j), match word[i:]
        def dfs(i, j, idx):
            if idx == len(word):
                return True

            if i < 0 or i >= m or j < 0 or j >= n or board[i][j] != word[idx]:
                return False
            
            tmp = board[i][j]
            board[i][j] = '#'

            found = (dfs(i + 1, j, idx + 1) or
                     dfs(i - 1, j, idx + 1) or
                     dfs(i, j + 1, idx + 1) or
                     dfs(i, j - 1, idx + 1)
                    )
            
            board[i][j] = tmp # backtrack
            return found


        for i in range(m):
            for j in range(n):
                # iterate each cell as start
                # and go through each char in word
                if dfs(i, j, 0):
                    return True
        
        return False
